server/src/main/java/org/elasticsearch/common/util/concurrent/ThrottledIterator.java

```diff
@@ -49,84 +58,49 @@ public static <T> void run(Iterator<T> iterator, BiConsumer<Releasable, T> itemC
     private final RefCounted refs; // one ref for each running item, plus one for the iterator if incomplete
     private final Iterator<T> iterator;
     private final BiConsumer<Releasable, T> itemConsumer;
-    private final Semaphore permits;
+    private final AtomicInteger permits;
+    private final Releasable itemReleasable = this::onItemRelease;
 
     private ThrottledIterator(Iterator<T> iterator, BiConsumer<Releasable, T> itemConsumer, int maxConcurrency, Runnable onCompletion) {
         this.iterator = Objects.requireNonNull(iterator);
         this.itemConsumer = Objects.requireNonNull(itemConsumer);
         if (maxConcurrency <= 0) {
             throw new IllegalArgumentException("maxConcurrency must be positive");
         }
-        this.permits = new Semaphore(maxConcurrency);
+        this.permits = new AtomicInteger(maxConcurrency);
         this.refs = AbstractRefCounted.of(onCompletion);
     }
 
     private void run() {
-        while (permits.tryAcquire()) {
+        do {
```

Comment:
I wonder if we can assert only one thread here at a time? Might be too close to the permits though to really make sense. Ok to skip if you do not find it useful.

At least adding a comment to `run()` that it is run in one thread at a time only would help reading this.